lab_3_keywords_textrank/main.py

```diff
@@ -824,7 +989,58 @@ def run(self) -> Optional[dict[str, dict[str, float]]]:
                 comparison report
         In case it is impossible to extract keywords due to corrupt inputs, None is returned
         """
-        pass
+        term_freq = {}
+        rake = {}
+        vanilla = {}
+        biased = {}
+
+        for index, theme in enumerate(self.themes):
+            with open(self._materials_path / f'{index}_text.txt', 'r', encoding='utf-8') as file:
+                text = file.read()
+            with open(self._materials_path / f'{index}_keywords.txt', 'r', encoding='utf-8') as file:
+                keywords = tuple(file.read().split())
+
+            preprocessor = TextPreprocessor(self._stop_words, self._punctuation)
+            clean_tokens = preprocessor.preprocess_text(text)
+            encoded_tokens = TextEncoder()
+            tokens = encoded_tokens.encode(clean_tokens)
+            if not tokens:
+                return None
+
+            tfidf_adapter = TFIDFAdapter(clean_tokens, self._idf)
+            tfidf_adapter.train()
+            term_freq[theme] = calculate_recall(tfidf_adapter.get_top_keywords(50), keywords)
+
+
+            rake_adapter = RAKEAdapter(text, self._stop_words)
+            rake_adapter.train()
+            rake[theme] = calculate_recall(rake_adapter.get_top_keywords(50), keywords)
+
+
+            graph = EdgeListGraph()
+            graph.fill_from_tokens(tokens, 3)
+            graph.fill_positions(tokens)
+            graph.calculate_position_weights()
+
+            vanilla_text_rank = VanillaTextRank(graph)
+            vanilla_text_rank.train()
+            decode_text_vanilla = encoded_tokens.decode(vanilla_text_rank.get_top_keywords(50))
+            if not decode_text_vanilla:
+                return None
+            vanilla[theme] = calculate_recall(decode_text_vanilla, keywords)
+
+            position_biased_text_rank = PositionBiasedTextRank(graph)
+            position_biased_text_rank.train()
```

Comment:
consider iterating over trainable objects instead of copying the pipeline
if you had aa million algorithms, would you also copy and paste?